### PR TITLE
Completed Lesson 04 & 05 Exercises & Assignments

### PR DESCRIPTION
**Pull request by trvvss**

### CODE REVIEW

**Comment by trvvss** [comment]
Completed Lesson 05 Exercise & Assignments as well. Please let me know if you see any opportunities for comprehensions in mailroom_part3.

File: students/travis_nelson/lesson05/mailroom_part3.py [hunk 1]
```diff
@@ -0,0 +1,163 @@
...
+
+def donation_response(donor):
+    """Returns the donation amount from the user"""
+    keep_asking = True
+    while keep_asking:
+        donation = input("How much did they donate?\n")
+        try:
+            if float(donation):
+                print("\nNice!\n")
+                keep_asking = False
```
Comment (natasha-aleksandrova) [inline students/travis_nelson/lesson05/mailroom_part3.py]: you don't actually need this bool for your loop - yes there is a good use for them in certain cases but not here.


`return donation` on line 70 (below) will already exit your function so you can just use `while True`

File: students/travis_nelson/lesson05/mailroom_part3.py [hunk 2]
```diff
@@ -0,0 +1,163 @@
...
+
+
+def sum_donations(donor_name):
+    """Returns sum of all historical donor donations"""
+    return round(sum(donor_db[donor_name]), 2)
+
+
+def sort_key(donor):
+    """Returns the key to sort donations by"""
+    return round(sum(donor_db[donor]))
```
Comment (natasha-aleksandrova) [inline students/travis_nelson/lesson05/mailroom_part3.py]: very small nitpick: it's good to have consistency 
`def sum_donations(donor_name):` and `def sort_key(donor):` have the same arguments that are named differently.

File: students/travis_nelson/lesson05/mailroom_part3.py [hunk 3]
```diff
@@ -0,0 +1,163 @@
...
+                    " donation of ${}.\n"
+                    "To date, you have donated ${} to our honorable mission.\n"
+                    "You are truly a valuable patron, and we thank you "
+                    "for your service (and money).\n"
+                    "{}\n"
+                    "{}".format(i,
+                                most_recent_donation,
+                                sum_donations(i),
+                                "Kindest regards,",
+                                "Baron Von Munchausen"))
```
Comment (natasha-aleksandrova) [inline students/travis_nelson/lesson05/mailroom_part3.py]: the two strings above don't need to be formatted with `format()`, you can use them inline with the string.

File: students/travis_nelson/lesson05/mailroom_part3.py [hunk 4]
```diff
@@ -0,0 +1,163 @@
...
+                        "Please choose from below options:\n",
+                        "1 - Send a Thank You to a single donor",
+                        "2 - Create a Report",
+                        "3 - Send letters to all donors",
+                        "4 - Quit\n\n",))
+    arg_dict = {"1": single_thank_you, "2": create_report,
+                "3": generate_all_thanks, "4": exit_program}
+    while True:
+        response = input(prompt)
+        arg_dict.get(response, invalid_response)()
```
Comment (natasha-aleksandrova) [inline students/travis_nelson/lesson05/mailroom_part3.py]: catch `KeyError` here

File: students/travis_nelson/lesson05/mailroom_part3.py [hunk 5]
```diff
@@ -0,0 +1,163 @@
...
+    keep_asking = True
+    while keep_asking:
+        donation = input("How much did they donate?\n")
+        try:
+            if float(donation):
+                print("\nNice!\n")
+                keep_asking = False
+                return donation
+        except ValueError:
+            print("\nJust a number is fine.\n")
```
Comment (natasha-aleksandrova) [inline students/travis_nelson/lesson05/mailroom_part3.py]: a bit more condensed versioN:
```
def donation_response(donor):
    while True:
        donation = input("How much did they donate?\n")
        try:
            return float(donation)
        except ValueError:
            print("\nJust a number is fine.\n")
```

File: students/travis_nelson/lesson05/mailroom_part3.py [hunk 6]
```diff
@@ -0,0 +1,163 @@
...
+          "To date, you have donated ${} to our honorable mission.\n"
+          "You are truly a valuable patron, and we thank you"
+          " for your service"
+          " (and money)".format(recipient, donation, total_donations))
+
+
+def recipient_response():
+    """Returns a recipient from the user...
+    If recipient was not previously in donor database, they are added"""
+    keep_asking = True
```
Comment (natasha-aleksandrova) [inline students/travis_nelson/lesson05/mailroom_part3.py]: I don't think you need this var as you never set it to False

File: students/travis_nelson/lesson05/mailroom_part3.py [hunk 7]
```diff
@@ -0,0 +1,163 @@
...
+                        print("\nAn existing patron!\n")
+                        return response
+                    else:
+                        print("\nA new donor!\n")
+                        donor_db[response] = []
+                        return response
+            else:
+                raise TypeError
+        except TypeError:
+            print("\nType a name, plz\n")
```
Comment (natasha-aleksandrova) [inline students/travis_nelson/lesson05/mailroom_part3.py]: are you anticipating any other line to throw `TypeError` besides line 56?

if not, then you might as well print your string in line 56

File: students/travis_nelson/lesson05/mailroom_part3.py [hunk 8]
```diff
@@ -0,0 +1,163 @@
...
+            if response[0].isalpha():
+                if response.lower() == "list":
+                    print("\n")
+                    for i in donor_db:
+                        print("{}".format(i))
+                else:
+                    donor_exists = False
+                    for donor in donor_db:
+                        if donor == response:
+                            donor_exists = True
```
Comment (natasha-aleksandrova) [inline students/travis_nelson/lesson05/mailroom_part3.py]: the huge benefit of using a dictionary for donors is ability to search efficiently by key:
```
if donor_db.get(response):
    return response
else:
    print('new donor')
```

**Pull request merged**